nixos/modules/services/networking/3proxy.nix

```diff
@@ -0,0 +1,427 @@
+    '');
+in with lib; {
+  options.services._3proxy = {
+    enable = mkEnableOption "3proxy";
+    confFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      example = "/var/lib/3proxy/3proxy.conf";
+      description = ''
+        Ignore all other 3proxy options and load configuration from this file.
+      '';
+    };
+    usersFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      example = "/var/lib/3proxy/3proxy.passwd";
+      description = ''
+        Load users and passwords from this file.
+
+        Example users file with plain-text passwords:
+
+        <para>
+          test1:CL:password1
+          test2:CL:password2
+        </para>
+
+        Example users file with md5-crypted passwords:
+
+        <para>
+          "test1:CR:$1$tFkisVd2$1GA8JXkRmTXdLDytM/i3a1"
+          "test2:CR:$1$rkpibm5J$Aq1.9VtYAn0JrqZ8M.1ME."
+        </para>
+
+        You can generate md5-crypted passwords via https://unix4lyfe.org/crypt/
+        Consult documentation at https://github.com/z3APA3A/3proxy/wiki/How-To-(incomplete)#USERS for more information.
+      '';
+    };
+    services = mkOption {
+      type = types.listOf (types.submodule {
+        options = {
+          type = mkOption {
+            type = types.enum [
+              "proxy"
+              "socks"
+              "pop3p"
+              "ftppr"
+              "admin"
+              "dnspr"
+              "tcppm"
+              "udppm"
+            ];
+            example = "proxy";
+            description = ''
+              Service type. The following values are valid:
+
+              <itemizedlist>
+                <listitem><para>
+                  <literal>"proxy"</literal>: HTTP/HTTPS proxy (default port 3128).
+                </para></listitem>
+                <listitem><para>
+                  <literal>"socks"</literal>: SOCKS 4/4.5/5 proxy (default port 1080).
+                </para></listitem>
+                <listitem><para>
+                  <literal>"pop3p"</literal>: POP3 proxy (default port 110).
+                </para></listitem>
+                <listitem><para>
+                  <literal>"ftppr"</literal>: FTP proxy (default port 21).
+                </para></listitem>
+                <listitem><para>
+                  <literal>"admin"</literal>: Web interface (default port 80).
+                </para></listitem>
+                <listitem><para>
+                  <literal>"dnspr"</literal>: Caching DNS proxy (default port 53).
+                </para></listitem>
+                <listitem><para>
+                  <literal>"tcppm"</literal>: TCP portmapper.
+                </para></listitem>
+                <listitem><para>
+                  <literal>"udppm"</literal>: UDP portmapper.
+                </para></listitem>
+              </itemizedlist>
+            '';
+          };
+          bindAddress = mkOption {
+            type = types.str;
+            default = "[::]";
```

Comment:
I am not sure if `[::]` will work when IPv6 is disabled, need to check
If not, then `default = if config.networking.enableIPv6 then "[::]" else "0.0.0.0"`

Reply:
It will. I added ```networking.enableIPv6 = false;``` to tests.